[TTI][AMDGPU] Allow targets to adjust `LastCallToStaticBonus` via `getInliningLastCallToStaticBonus`

shiltian opened the pull request:
Currently we will not be able to inline a large function even if it only has one live use because the inline cost is still very high after applying `LastCallToStaticBonus`, which is a constant. This could significantly impact the performance because CSR spill is very expensive.

This PR adds a new function `getInliningLastCallToStaticBonus` to TTI to allow targets to customize this value.

Fixes SWDEV-471398.

shiltian commented:
* **#111311** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/111311?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

jmmartinez commented:
> but the cost is still very close to the threshold: cost=14010, threshold=170775.

It's a 10x difference. Looks pretty safe to me. What am I missing?

TBH I do not see much of an easy alternative to what you're doing.

---

To test, instead of using big functions, you could pass `-inline-threshold=0 -debug-only=inline-cost` and check that the inline threshold increases between a function that is called once and another that is called twice.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1307,6 +1311,12 @@ unsigned GCNTTIImpl::adjustInliningThreshold(const CallBase *CB) const {
+
+  // Increase the threshold if it is the only call to a local function.
+  Function *Callee = CB->getCalledFunction();
+  if (Callee->hasLocalLinkage() && Callee->hasOneLiveUse())
```
Guard against null callee

arsenm reviewed (commented):
Needs test. Isn't there a generic control for this already?

nikic reviewed (commented):
I'm a bit confused here. This already exists in the form of a cost bonus in the generic inlining cost model.

Is the point here that you want to effectively double the bonus by both applying it as a bonus in the generic model and as a threshold adjustment in AMDGPU TTI?

shiltian commented:
> I'm a bit confused here. This already exists in the form of a cost bonus in the generic inlining cost model.
> 
> Is the point here that you want to effectively double the bonus by both applying it as a bonus in the generic model and as a threshold adjustment in AMDGPU TTI?

Yes. IIUC, the generic model doesn't seem to have a target dependent approach to adjust the bonus. The default value is not sufficient and we want to inline it regardless.

arsenm commented:
> Yes. IIUC, the generic model doesn't seem to have a target dependent approach to adjust the bonus. The default value is not sufficient and we want to inline it regardless.

I'd rather avoid splitting the logic for this. Where is the default handling?

shiltian commented:
> Where is the default handling?

https://github.com/llvm/llvm-project/blob/8a9e9a89f0f5d58d1dacf3a7e626d8eb1c4ea485/llvm/lib/Analysis/InlineCost.cpp#L2032

shiltian commented:
> It's a 10x difference. Looks pretty safe to me. What am I missing?

I missed one figure Lol.

shiltian commented:
Sure. I can do that.

shiltian commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -1943,7 +1943,7 @@ void InlineCostCallAnalyzer::updateThreshold(CallBase &Call, Function &Callee) {
   // and the callsite.
   int SingleBBBonusPercent = 50;
   int VectorBonusPercent = TTI.getInlinerVectorBonusPercent();
-  int LastCallToStaticBonus = InlineConstants::LastCallToStaticBonus;
```
We still can't delete `InlineConstants::LastCallToStaticBonus` because it is used in another place where there is no access to TTI (including its parent function).

https://github.com/llvm/llvm-project/blob/4169338e75cdce73d34063532db598c95ee82ae4/llvm/lib/Analysis/InlineAdvisor.cpp#L323

shiltian commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -352,6 +352,9 @@ class TargetTransformInfo {
   unsigned getInliningCostBenefitAnalysisProfitableMultiplier() const;
 
+  /// \returns The bonus of inlining the last call to a static function.
+  int getInliningLastCallToStaticBonus() const;
```
I'm not sure if it makes sense to use `unsigned` as return value, but `InlineConstants::LastCallToStaticBonus` is `int`.

shiltian replied:
Also, not sure if it is better to make it as `getInliningLastCallToStaticBonus(const CallBase *CB)`.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -75,6 +75,10 @@ static cl::opt<size_t> InlineMaxBB(
              " (compile time constraint)"));
 
+static cl::opt<unsigned> InlineThresholdOneLiveUse(
+    "amdgpu-inline-threshold-one-live-use", cl::Hidden, cl::init(165000),
```
165000 is the default value 15000 multiplied by AMDGPU's inlining threshold multiplier`, which is 11.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -75,6 +75,10 @@ static cl::opt<size_t> InlineMaxBB(
              " (compile time constraint)"));
 
+static cl::opt<unsigned> InlineLastCallToStaticBonus(
+    "amdgpu-inline-last-call-to-static-bonus", cl::Hidden, cl::init(165000),
```
165000 is the default value 15000 multiplied by AMDGPU's inlining threshold multiplier, which is 11.

jmmartinez replied:
What do you think about keeping it as a constant and not an option?

I don't think nobody is ever going to use that option.

And in your test the bonus contribution is already clear by comparing the scores for `@callee_not_only_one_live_use` and `@callee_only_one_live_use`.

shiltian replied:
> What do you think about keeping it as a constant and not an option?

I don't have strong opinions on it. I'm fine with making it a constant.

arsenm commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -74,6 +74,10 @@ class TargetTransformInfoImplBase {
     return 8;
   }
+  int getInliningLastCallToStaticBonus() const {
+    // This is same as InlineConstants::LastCallToStaticBonus.
+    return 15000;
```
Just use LastCallToStaticBonus instead of hoping it stays the same?

shiltian replied:
I don't want to include `llvm/Analysis/InlineCost.h` here because here it is a header as well, but apparently forward declaration doesn't work because the `InlineConstants::LastCallToStaticBonus` defined in a header as well. Any suggestions?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h:
```diff
@@ -240,6 +240,11 @@ class GCNTTIImpl final : public BasicTTIImplBase<GCNTTIImpl> {
                            const Function *Callee) const;
 
+  int getInliningLastCallToStaticBonus() const {
+    // InlineConstants::LastCallToStaticBonus * 11 where 11 is AMDGPU's inlining
+    // threshold multiplier.
+    return 165000;
```
Can you just take the existing multiplier and apply it to the existing threshold without the new layer? I would have guessed the multiplier applying to bonuses and costs would cancel out

shiltian replied:
> Can you just take the existing multiplier and apply it to the existing threshold without the new layer?

That is what currently is doing but apparently the threshold is not large enough for a large function even after the constant bonus applied.

arsenm replied:
The multiply is not applied to the LastCallToStaticBonus below

shiltian replied:
That is gonna change all the targets, which I'm not sure if that's a good idea. That is also the motivation to have a target dependent API instead of assuming all targets want to inline the last call.

arsenm replied:
It makes no sense to have this multiplier applied to some places and not the others. Do not do things out of fear of touching other targets, that's how we end up with a lot of the messes we already have

shiltian replied:
I applied the multiplier just by choice. It is not like the general model should apply it. I can just set a random large number. There is no general rule (or studied heuristic) here, which exactly echoes the need for a target dependent bonus.

arsenm replied:
Just like the threshold multiplier hook we already have. It's basically the do GPU inlining flag already

shiltian replied:
So you were suggesting to add a bonus multiplier?

arsenm replied:
I mean it's preferable to just implement this in terms of what we already have. If the multiplier is already applied to other bonuses, and not this bonus, it's easier and more consistent to just apply it there too.

shiltian replied:
I don't think the multiplier was already applied to other bonuses. For example, `VectorBonus` has a TTI function `getInlinerVectorBonusPercent` to determine the "multiplier".

arsenm commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -74,6 +74,11 @@ class TargetTransformInfoImplBase {
     return 8;
   }
+  int getInliningLastCallToStaticBonus() const {
+    // This is the value of InlineConstants::LastCallToStaticBonus before this
+    // function was introduced.
+    return 15000;
```
Can you just use the constant enum instead of hardcoding it with comment?

shiltian replied:
I removed the constant in this PR.

shiltian commented on llvm/include/llvm/Analysis/InlineCost.h:
```diff
@@ -47,7 +47,6 @@ const int OptAggressiveThreshold = 250;
 int getInstrCost();
 const int IndirectCallThreshold = 100;
 const int LoopPenalty = 25;
-const int LastCallToStaticBonus = 15000;
```
@arsenm It's here

shiltian commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -74,6 +74,11 @@ class TargetTransformInfoImplBase {
     return 8;
   }
+  int getInliningLastCallToStaticBonus() const {
+    // This is the value of InlineConstants::LastCallToStaticBonus before this
```
I just wanted to mention here that this is not a random value. It is the previous `InlineConstants::LastCallToStaticBonus` which got removed in this PR.

shiltian committed:
[AMDGPU] Increase inline threshold when the callee only has one live use

Currently we will not inline a large function even if it only has one live use.
This could significantly impact the performance because CSR spill is very
expensive. The goal of this PR is trying to force the inlining if there is only
one live use by adjusting the inlining threshold, which is a configurable
number. The default value is 15000, which borrows from
`InlineConstants::LastCallToStaticBonus`. I'm not sure if this is a good number,
and if this is the right way to do that. After making this change, the callee in
my local test case can finally be inlined, but the cost is still very close to
the threshold: `cost=14010, threshold=170775`.

Speaking of the test, how are we gonna test this? Do we want to include a giant
IR file?

Fixes SWDEV-471398.
